File: apps/apple/Sources/BeeboKit/TVLink.swift
import Foundation

public struct TVRequest: Equatable, Sendable {
    public let deviceName: String
    public let deviceModel: String
    public let requestedMinutesAgo: Int
    public let expiresIn: Int

    public init(deviceName: String, deviceModel: String, requestedMinutesAgo: Int, expiresIn: Int) {
        self.deviceName = deviceName
        self.deviceModel = deviceModel
        self.requestedMinutesAgo = requestedMinutesAgo
        self.expiresIn = expiresIn
    }
}

public enum LinkError: Equatable, Sendable {
    case invalidCode
    case rateLimited
    case unauthorized
    case notAllowed
    case noHome
    case passwordReset
    case accountInactive
    case invalidCredentials
    case unavailable
    case offline
    case server
}

public enum LinkResult<Value: Sendable>: Sendable {
    case ok(Value)
    case refused(LinkError, retryAfter: Int = 0)
}

extension LinkResult: Equatable where Value: Equatable {}

public enum TVDecision: String, Sendable {
    case approve
    case deny
}

public protocol TVLinkService: Sendable {
    func lookup(token: String, userCode: String) async -> LinkResult<TVRequest>
    func decide(token: String, userCode: String, decision: TVDecision) async -> LinkResult<TVRequest>
}

public enum FindHomeResult: Equatable, Sendable {
    case found(name: String, token: String)
    case failed(LinkError)
}

public enum TVLinkParsing {
    private static func object(_ data: Data) -> [String: Any]? {
        try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func int(_ value: Any?) -> Int? {
        if let n = value as? Int { return n }
        if let d = value as? Double { return Int(d) }
        return nil
    }

    static func refusal(status: Int, body: [String: Any]?) -> LinkResult<TVRequest> {
        let error = body?["error"] as? String
        switch true {
        case status == 429:
            return .refused(.rateLimited, retryAfter: min(3600, max(0, int(body?["retry_after"]) ?? 0)))
        case status == 401:
            return .refused(.unauthorized)
        case error == "not_allowed":
            return .refused(.notAllowed)
        case error == "no_home":
            return .refused(.noHome)
        case error == "password_reset_required":
            return .refused(.passwordReset)
        case error == "invalid_code":
            return .refused(.invalidCode)
        case status == 404:
            return .refused(.unavailable)
        default:
            return .refused(.server)
        }
    }

    private static func request(_ body: [String: Any]) -> TVRequest {
        let name = (body["device_name"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "TV"
        return TVRequest(
            deviceName: name,
            deviceModel: (body["device_model"] as? String) ?? "",
            requestedMinutesAgo: max(0, int(body["requested_minutes_ago"]) ?? 0),
            expiresIn: max(0, int(body["expires_in"]) ?? 0)
        )
    }

    public static func parseLookup(status: Int, data: Data) -> LinkResult<TVRequest> {
        let body = object(data)
        if status == 200, let body, body["device_name"] != nil { return .ok(request(body)) }
        return refusal(status: status, body: body)
    }

    public static func parseDecision(status: Int, data: Data, decision: TVDecision) -> LinkResult<TVRequest> {
        let body = object(data)
        let expected = decision == .approve ? "approved" : "denied"
        if status == 200, let body, (body["status"] as? String) == expected { return .ok(request(body)) }
        return refusal(status: status, body: body)
    }

    public static func parseFindHome(status: Int, data: Data) -> FindHomeResult {
        let body = object(data)
        if status == 200, let body, let name = body["name"] as? String, !name.isEmpty, let token = body["token"] as? String, !token.isEmpty {
            return .found(name: name, token: token)
        }
        switch status {
        case 401: return .failed(.invalidCredentials)
        case 402: return .failed(.accountInactive)
        case 429: return .failed(.rateLimited)
        case 404: return .failed(.unavailable)
        default: return .failed(.server)
        }
    }
}

public enum TVLinkMessages {
    public static func message(for error: LinkError, retryAfter: Int = 0) -> String {
        switch error {
        case .invalidCode:
            return "That code isn't right, or it has run out. Check the code on your TV. A new one appears every few minutes."
        case .rateLimited:
            let minutes = max(1, (retryAfter + 59) / 60)
            return "Too many tries. Wait \(minutes) minute\(minutes == 1 ? "" : "s") and try again."
        case .unauthorized:
            return "Your Beebo account sign-in didn't work. Sign in again, then try once more."
        case .notAllowed:
            return "This sign-in can't link a TV. Sign in with your own Beebo account."
        case .noHome:
            return "This account doesn't have a Beebo home yet, so there is nothing for a TV to connect to."
        case .passwordReset:
            return "This account needs a new password first."
        case .accountInactive:
            return "This Beebo account isn't active. Manage your account at beebo.tv."
        case .invalidCredentials:
            return "That email or password isn't right."
        case .unavailable:
            return "Linking a TV isn't available yet."
        case .offline:
            return "Couldn't reach Beebo. Check your connection and try again."
        case .server:
            return "Beebo didn't answer properly. Try again in a moment."
        }
    }

    public static func ago(minutes: Int) -> String {
        switch minutes {
        case ...0: return "just now"
        case 1: return "1 minute ago"
        default: return "\(minutes) minutes ago"
        }
    }
}

File: apps/core/app/src/main/java/com/beeboentertainment/movie/tvpair/TvLink.kt
package com.beeboentertainment.movie.tvpair

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull

/*
 * The phone's side: type the code a TV is showing, see which TV asked, approve or deny.
 * Same rules as TvPairing.kt: no Android types here.
 */

/** What the phone shows about the TV that asked, so the person can recognise it. */
data class TvRequest(
    val deviceName: String,
    val deviceModel: String,
    val requestedMinutesAgo: Int,
    val expiresInS: Int,
)

enum class LinkError {
    /** Unknown, expired or already used. The Worker says the same for all three. */
    INVALID_CODE,
    /** Too many wrong codes from this phone or account. */
    RATE_LIMITED,
    /** This phone holds no valid Beebo account sign-in. */
    UNAUTHORIZED,
    /** A guest or household-pass sign-in: not a person. */
    NOT_ALLOWED,
    /** The account has no Beebo home for the TV to connect to. */
    NO_HOME,
    PASSWORD_RESET,
    UNAVAILABLE,
    OFFLINE,
    SERVER,
}

sealed interface LinkResult<out T> {
    data class Ok<T>(val value: T) : LinkResult<T>
    data class Refused(val error: LinkError, val retryAfterS: Int = 0) : LinkResult<Nothing>
}

enum class TvDecision(val wire: String) { APPROVE("approve"), DENY("deny") }

object TvLinkParsing {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    private fun obj(body: String): JsonObject? = runCatching { json.parseToJsonElement(body) as? JsonObject }.getOrNull()
    private fun JsonObject.str(k: String): String? = (this[k] as? JsonPrimitive)?.contentOrNull
    private fun JsonObject.int(k: String): Int? = (this[k] as? JsonPrimitive)?.intOrNull

    private fun refusal(httpCode: Int, o: JsonObject?): LinkResult.Refused {
        val error = o?.str("error")
        return when {
            httpCode == 429 -> LinkResult.Refused(LinkError.RATE_LIMITED, (o?.int("retry_after") ?: 0).coerceIn(0, 3600))
            httpCode == 401 -> LinkResult.Refused(LinkError.UNAUTHORIZED)
            error == "not_allowed" -> LinkResult.Refused(LinkError.NOT_ALLOWED)
            error == "no_home" -> LinkResult.Refused(LinkError.NO_HOME)
            error == "password_reset_required" -> LinkResult.Refused(LinkError.PASSWORD_RESET)
            error == "invalid_code" -> LinkResult.Refused(LinkError.INVALID_CODE)
            httpCode == 404 -> LinkResult.Refused(LinkError.UNAVAILABLE)
            else -> LinkResult.Refused(LinkError.SERVER)
        }
    }

    private fun request(o: JsonObject) = TvRequest(
        deviceName = o.str("device_name")?.ifBlank { null } ?: "TV",
        deviceModel = o.str("device_model").orEmpty(),
        requestedMinutesAgo = (o.int("requested_minutes_ago") ?: 0).coerceAtLeast(0),
        expiresInS = (o.int("expires_in") ?: 0).coerceAtLeast(0),
    )

    /** Answer to POST /tvpair/lookup. */
    fun parseLookup(httpCode: Int, body: String): LinkResult<TvRequest> {
        val o = obj(body)
        if (httpCode == 200 && o != null && o.str("device_name") != null) return LinkResult.Ok(request(o))
        return refusal(httpCode, o)
    }

    /** Answer to POST /tvpair/approve with a decision. */
    fun parseDecision(httpCode: Int, body: String, decision: TvDecision): LinkResult<TvRequest> {
        val o = obj(body)
        if (httpCode == 200 && o != null && o.str("status") == (if (decision == TvDecision.APPROVE) "approved" else "denied")) {
            return LinkResult.Ok(request(o))
        }
        return refusal(httpCode, o)
    }
}

object TvLinkMessages {
    fun forError(error: LinkError, retryAfterS: Int = 0): String = when (error) {
        LinkError.INVALID_CODE -> "That code isn't right, or it has run out. Check the code on your TV. A new one appears every few minutes."
        LinkError.RATE_LIMITED -> {
            val m = ((retryAfterS + 59) / 60).coerceAtLeast(1)
            "Too many tries. Wait $m minute${if (m == 1) "" else "s"} and try again."
        }
        LinkError.UNAUTHORIZED -> "This phone isn't signed in to your Beebo account right now. Sign in again, then try once more."
        LinkError.NOT_ALLOWED -> "This sign-in can't link a TV. Sign in with your own Beebo account."
        LinkError.NO_HOME -> "This account doesn't have a Beebo home yet, so there is nothing for a TV to connect to."
        LinkError.PASSWORD_RESET -> "This account needs a new password first."
        LinkError.UNAVAILABLE -> "Linking a TV isn't available yet."
        LinkError.OFFLINE -> "Couldn't reach Beebo. Check your connection and try again."
        LinkError.SERVER -> "Beebo didn't answer properly. Try again in a moment."
    }

    /** "3 minutes ago", coarse on purpose: enough to know it is the TV just switched on. */
    fun ago(minutes: Int): String = when {
        minutes <= 0 -> "just now"
        minutes == 1 -> "1 minute ago"
        else -> "$minutes minutes ago"
    }
}

/** The phone side of the protocol. */
interface TvLinkService {
    suspend fun lookup(token: String, userCode: String): LinkResult<TvRequest>
    suspend fun decide(token: String, userCode: String, decision: TvDecision): LinkResult<TvRequest>
}

/**
 * A code arriving from outside the screen (the beebo://tv-link deep link). Held until the signed-in
 * app is showing, then consumed by the Link a TV screen.
 */
object TvLinkRequests {
    private val _pending = MutableStateFlow<String?>(null)
    val pending: StateFlow<String?> = _pending.asStateFlow()

    /** [dataString] is the intent's data. Returns whether it was a link to this screen. */
    fun offer(dataString: String?): Boolean {
        val s = dataString?.trim().orEmpty()
        if (!s.startsWith(SCHEME_PREFIX, ignoreCase = true)) return false
        _pending.value = TvPairCodes.fromLinkOrText(s) ?: ""
        return true
    }

    fun consume() { _pending.value = null }

    const val SCHEME_PREFIX = "beebo://tv-link"
}
